// nx/src/utils/ab-testing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.recordStat = exports.messages = exports.PromptMessages = void 0;
const tslib_1 = require("tslib");
const axios_1 = require("axios");
const is_ci_1 = require("./is-ci");
class PromptMessages {
    constructor() {
        this.messages = {
            nxCloudCreation: [
                {
                    code: 'set-up-distributed-caching-ci',
                    message: `Enable distributed caching to make your CI faster`,
                },
            ],
            nxCloudMigration: [
                {
                    code: 'we-noticed',
                    message: `We noticed you are migrating to a new major version, but are not taking advantage of Nx Cloud. Nx Cloud can make your CI up to 10 times faster. Learn more about it here: nx.app. Would you like to add it?`,
                },
                {
                    code: 'not-leveraging-caching',
                    message: `You're not leveraging distributed caching yet. Do you want to enable it and speed up your CI?`,
                },
                {
                    code: 'make-ci-faster',
                    message: `Enable distributed caching to make your CI faster?`,
                },
            ],
        };
        this.selectedMessages = {};
    }
    getPromptMessage(key) {
        if (this.selectedMessages[key] === undefined) {
            if (process.env.NX_GENERATE_DOCS_PROCESS === 'true') {
                this.selectedMessages[key] = 0;
            }
            else {
                this.selectedMessages[key] = Math.floor(Math.random() * this.messages[key].length);
            }
        }
        return this.messages[key][this.selectedMessages[key]].message;
    }
    codeOfSelectedPromptMessage(key) {
        if (this.selectedMessages[key] === undefined)
            return null;
        return this.messages[key][this.selectedMessages[key]].code;
    }
}
exports.PromptMessages = PromptMessages;
exports.messages = new PromptMessages();
/**
 * We are incrementing a counter to track how often create-nx-workspace is used in CI
 * vs dev environments. No personal information is collected.
 */
function recordStat(opts) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        try {
            const major = Number(opts.nxVersion.split('.')[0]);
            if (process.env.NX_VERBOSE_LOGGING === 'true') {
                console.log(`Record stat. Major: ${major}`);
            }
            if (major < 10 || major > 14)
                return; // test version, skip it
            yield axios_1.default
                .create({
                baseURL: 'https://cloud.nx.app',
                timeout: 400,
            })
                .post('/nx-cloud/stats', {
                command: opts.command,
                isCI: (0, is_ci_1.isCI)(),
                useCloud: opts.useCloud,
                meta: opts.meta,
            });
        }
        catch (e) {
            if (process.env.NX_VERBOSE_LOGGING === 'true') {
                console.error(e);
            }
        }
    });
}
exports.recordStat = recordStat;
